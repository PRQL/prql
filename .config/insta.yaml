behavior:
  # force_update: true
review:
  # The default (true) has a small performance hit and never needed for us
  warn_undiscovered: false
  include_ignored: false
  include_hidden: false
