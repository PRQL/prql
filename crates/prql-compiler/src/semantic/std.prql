# The PRQL standard library defines the following functions and transforms.
# The definitions are whitespace insensitive, and have this form:
#
# ```
# let my_func = param1 param2 ...  -> <return_type> body_expr
# ```
#
# Where:
#   * `my_func` is the name of the function
#   * `param1` is the first parameter optionally followed by a type in "< ... >"
#   * `param2` etc. follow the same pattern as param1
#   * `<return_type>` is the type of result wrapped in "< ... >"
#   * `body_expr` defines the function body that creates the result.
#     It can be PRQL code or `internal ...` to indicate internal compiler code.

# Operators

let mul = left right -> <int || float> internal std.mul
let div_i = left right -> <int || float> internal std.div_i
let div_f = left right -> <int || float> internal std.div_f
let mod = left right -> <int || float> internal std.mod
let add = left<int || float || timestamp || date> right<int || float || timestamp || date> -> <int || float || timestamp || date> internal std.add
let sub = left<int || float || timestamp || date> right<int || float || timestamp || date> -> <int || float || timestamp || date> internal std.sub
let eq = left right -> <bool> internal std.eq
let ne = left right -> <bool> internal std.ne
let gt = left right -> <bool> internal std.gt
let lt = left right -> <bool> internal std.lt
let gte = left right -> <bool> internal std.gte
let lte = left right -> <bool> internal std.lte
let and = left<bool> right<bool> -> <bool> internal std.and
let or = left<bool> right<bool> -> <bool> internal std.or
let coalesce = left right -> internal std.coalesce
let regex_search = text pattern -> <bool> internal std.regex_search

let neg = expr<int || float> -> <int || float> internal std.neg
let not = expr<bool> -> <bool> internal std.not

# Types

## Type primitives
type int
type float
type bool
type text
type date
type time
type timestamp
type `func`

## Generic array
# TODO: an array of anything, not just nulls
type array = [null]

## Scalar
type scalar = int || float || bool || text || date || time || timestamp || null
type tuple_of_scalars = {scalar..}

## Relation (an array of tuples)
type relation = [tuple_of_scalars]

## Transform
type transform = (func relation -> relation)

# Functions

## Relational transforms
let from = func
  `default_db.source` <relation>
  -> <relation> internal from

let select = func
  columns <scalar || tuple_of_scalars>
  tbl <relation>
  -> <relation> internal select

let filter = func
  condition <bool>
  tbl <relation>
  -> <relation> internal filter

let derive = func
  columns <scalar || tuple_of_scalars>
  tbl <relation>
  -> <relation> internal derive

let aggregate = func
  columns <scalar || tuple_of_scalars>
  tbl <relation>
  -> <relation> internal aggregate

let sort = func
  by <scalar || tuple_of_scalars>
  tbl <relation>
  -> <relation> internal sort

let take = func
  expr <scalar>
  tbl <relation>
  -> <relation> internal take

let join = func
  `default_db.with` <relation>
  condition <bool>
  `noresolve.side`:inner
  tbl <relation>
  -> <relation> internal join

let group = func
  by<scalar || tuple_of_scalars>
  pipeline <transform>
  tbl <relation>
  -> <relation> internal group

let window = func
  rows:0..0
  range:0..0
  expanding <bool>:false
  rolling <int>:0
  pipeline <transform>
  tbl <relation>
  -> <relation> internal window

let noop = x -> x

let append = `default_db.bottom`<relation> top<relation> -> <relation> internal append
let intersect = `default_db.bottom`<relation> top<relation> -> <relation> (
  noop t = top
  join (noop b = bottom) (tuple_every (tuple_map _eq (tuple_zip t.* b.*)))
  select t.*
)
let remove = `default_db.bottom`<relation> top<relation> -> <relation> (
  noop t = top
  join side:left (noop b = bottom) (tuple_every (tuple_map _eq (tuple_zip t.* b.*)))
  filter (tuple_every (tuple_map _is_null b.*))
  select t.*
)
let loop = func
  pipeline <transform>
  top <relation>
  -> <relation> internal loop

## Aggregate functions
# These return either a scalar when used within `aggregate`, or a column when used anywhere else.

let min = column <array> -> <int || float || null> internal std.min

let max = column <array> -> <int || float || null> internal std.max

let sum = column <array> -> <int || float> internal std.sum

let average = column <array> -> <float || null> internal std.average

let stddev = column <array> -> <float || null> internal std.stddev

let every = column <array> -> <bool> internal std.every

let any = column <array> -> <bool> internal std.any

let concat_array = column <array> -> <text> internal std.concat_array

# Counts number of items in the column.
# Note that the count will include null values.
let count = column<array> -> <int> internal std.count

# Deprecated in favour of filterning input to the [std.count] function (not yet implemented).
@{deprecated}
let count_distinct = column <array> -> internal std.count_distinct

## Window functions
let lag =   offset <int>    column <array> -> internal std.lag
let lead =  offset <int>    column <array> -> internal std.lead
let first      = column <array> -> internal std.first
let last       = column <array> -> internal std.last
let rank       = column <array> -> internal std.rank
let rank_dense = column <array> -> internal std.rank_dense
let row_number = column <array> -> internal std.row_number

## Misc functions
let round = n_digits column -> <scalar> internal std.round
let as = `noresolve.type` column -> <scalar> internal std.as
let in = pattern value -> <bool> internal in

## Tuple functions
let tuple_every = func list -> <bool> internal tuple_every
let tuple_map = func fn <func> list -> internal tuple_map
let tuple_zip = func a b -> internal tuple_zip
let _eq = func a -> internal _eq
let _is_null = func a -> _param.a == null

## Misc
let from_text = input<text> `noresolve.format`:csv -> <relation> internal from_text

## String functions
let lower = column -> <text> internal std.lower
let upper = column -> <text> internal std.upper

## File-reading functions, primarily for DuckDB
let read_parquet = source<text> -> <relation> internal std.read_parquet
let read_csv = source<text> -> <relation> internal std.read_csv
