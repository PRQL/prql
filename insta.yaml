behavior:
  force_update: true
review:
  include_ignored: false
  include_hidden: false
  # Small performance hit and never needed for us
  warn_undiscovered: false
