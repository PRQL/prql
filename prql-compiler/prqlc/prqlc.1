.\" prqlc man page.
.\"
.\" PRQL is a modern language for transforming data — a simple, powerful,
.\" pipelined SQL replacement.
.\"
.\" https://prql-lang.org/
.Dt prqlc 1
.Os Linux
.Sh NAME
.Nm prqlc
.And PRQL compiler.
.Sh SYNOPSIS
.Pp
.SY prqlc
parse
.Op Ar input
.Op Ar output
.YS
.Pp
.SY prqlc
fmt
.Op Ar input
.Op Ar output
.YS
.Pp
.SY prqlc
annotate
.Op Ar input
.Op Ar output
.YS
.Pp
.SY prqlc
debug
.Op Ar input
.Op Ar output
.YS
.Pp
.SY prqlc
resolve
.Op Ar input
.Op Ar output
.YS
.Pp
.SY prqlc
compile
.Op Ar options
.Op Ar input
.Op Ar output
.YS
.Pp
.SY prqlc
watch
.Op Ar options
path
.YS
.Pp
.SY prqlc
help
.YS
.Pp
.Sh DESCRIPTION
PRQL is a modern language for transforming data — a simple, powerful,
pipelined SQL replacement.
.Pp
Flags:
.Bl -tag
.It Fl h
Print help
.It Fl v
Print version
.El
.Pp
Flags for the compile option:
.Bl -tag
.It Fl -no-format
do not format the output
.It Fl -no-signature
do not print the PRQL signature
.El
.Sh FILES                \" File used or created by the topic of the man page
.Bl -tag
.It Pa /usr/bin/prqlc
The
.Nm
executable.
.Sh DIAGNOSTICS
.Nm
returns zero on normal operation, 1 on error.
.Sh BUGS
Report any  bugs on https://github.com/PRQL/prql/issues
.Sh AUTHOR
PRQL team
