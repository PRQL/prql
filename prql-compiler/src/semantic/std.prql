# Aggregate Functions
func min <scalar|column> column -> null
func max <scalar|column> column -> null
func sum <scalar|column> column -> null
func avg <scalar|column> column -> null
func stddev <scalar|column> column -> null
func average <scalar|column> column -> null
func count <scalar|column> non_null:s"*" -> null
# TODO: Possibly make this into `count distinct:true` (or like `distinct:` as an
# abbreviation of that?)
func count_distinct <scalar|column> column -> null

# Window functions
func lag<column> offset column -> null
func lead<column> offset column -> null
func first<column> offset column -> null
func last<column> offset column -> null
func rank<column> -> null
func rank_dense<column> -> null
func row_number<column> -> null

# Other functions
func round<scalar> n_digits column -> null
func as<scalar> `noresolve.type` column -> null
func in<bool> pattern value -> null

# Transform type definitions
func from<table> `default_db.source`<table> -> null
func select<table> columns<column> tbl<table> -> null
func filter<table> condition<bool> tbl<table> -> null
func derive<table> columns<column> tbl<table> -> null
func aggregate<table> a<column> tbl<table> -> null
func sort<table> by tbl<table> -> null
func take<table> expr tbl<table> -> null
func join<table> `default_db.with`<table> filter `noresolve.side`:inner tbl<table> -> null
func group<table> by pipeline tbl<table> -> null
func window<table> rows:0..0 range:0..0 expanding:false rolling:0 pipeline tbl<table> -> null

func append<table> `default_db.bottom`<table> top<table> -> null
func intersect<table> `default_db.bottom`<table> top<table> -> (
    from t = _param.top
    join b = _param.bottom (all (map _eq (zip t.* b.*)))
    select t.*
)
func remove<table> `default_db.bottom`<table> top<table> -> (
    from t = _param.top
    join side:left b = _param.bottom (all (map _eq (zip t.* b.*)))
    filter (all (map _is_null b.*))
    select t.*
)
func loop<table> pipeline top<table> -> null

# List functions
func all<bool> list<list> -> null
func map<list> fn list<list> -> null
func zip<list> a<list> b<list> -> null
func _eq<list> a<list> -> null
func _is_null a -> _param.a == null

# Misc
func from_text<table> text<string> `noresolve.format`:csv -> null
