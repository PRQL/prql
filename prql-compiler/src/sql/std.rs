use std::collections::HashMap;
use std::iter::zip;

use anyhow::Result;
use once_cell::sync::Lazy;
use sqlparser::ast::{self as sql_ast};

use super::gen_expr::translate_sstring;
use super::Context;
use crate::ast::{pl, rq};
use crate::semantic;

static STD: Lazy<semantic::Module> = Lazy::new(load_std_impl);

fn load_std_impl() -> semantic::Module {
    use crate::parser::parse;
    let std_lib = include_str!("./std_impl.prql");
    let statements = parse(std_lib).unwrap();

    let context = semantic::Context {
        root_mod: semantic::Module::new(),
        ..semantic::Context::default()
    };

    let (_, context) = semantic::resolve_only(statements, Some(context)).unwrap();
    let std = context.root_mod.get(&pl::Ident::from_name("std")).unwrap();

    std.kind.clone().into_module().unwrap()
}

pub(super) fn translate_built_in(
    name: String,
    args: Vec<rq::Expr>,
    ctx: &mut Context,
) -> Result<sql_ast::Expr> {
    let name = name.strip_prefix("std.").unwrap();

    let entry = STD.get(&pl::Ident::from_name(name)).unwrap();
    let func_def = entry.kind.as_func_def().unwrap();

    let params = func_def
        .named_params
        .iter()
        .chain(func_def.positional_params.iter())
        .map(|x| x.name.split('.').last().unwrap_or(x.name.as_str()));

    let mut args: HashMap<&str, _> = zip(params, args.into_iter()).collect();

    // body can only be an s-string
    let body = &func_def.body.kind.as_s_string().unwrap();
    let body = body
        .iter()
        .map(|item| {
            match item {
                pl::InterpolateItem::Expr(expr) => {
                    // s-string exprs can only contain idents
                    let ident = expr.kind.as_ident();
                    let ident = ident.as_ref().unwrap();

                    // lookup args
                    let arg = args.remove(ident.name.as_str());
                    pl::InterpolateItem::<rq::Expr>::Expr(Box::new(arg.unwrap()))
                }
                pl::InterpolateItem::String(s) => pl::InterpolateItem::String(s.clone()),
            }
        })
        .collect::<Vec<_>>();

    let s_string = translate_sstring(body, ctx)?;

    Ok(sql_ast::Expr::Identifier(sql_ast::Ident::new(s_string)))
}
