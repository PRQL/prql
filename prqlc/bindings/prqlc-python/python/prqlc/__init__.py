from .prqlc import *

__doc__ = prqlc.__doc__
if hasattr(prqlc, "__all__"):
    __all__ = prqlc.__all__