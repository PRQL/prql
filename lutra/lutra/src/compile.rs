use std::collections::HashMap;
use std::str::FromStr;

use anyhow::Result;
use prqlc::ir::decl::RootModule;
use prqlc::ir::pl::{Ident, Literal};
use prqlc::sql::Dialect;
use prqlc::{semantic, ErrorMessages};
use prqlc::{Error, Errors, Options, SourceTree, Target, WithErrorInfo};

use crate::project::{DatabaseModule, ProjectCompiled, ProjectDiscovered, SqliteConnectionParams};

#[cfg_attr(feature = "clap", derive(clap::Parser))]
#[derive(Default)]
pub struct CompileParams {}

pub fn compile(mut project: ProjectDiscovered, _: CompileParams) -> Result<ProjectCompiled> {
    let files = std::mem::take(&mut project.sources);
    let source_tree = SourceTree::new(files, Some(project.root_path.clone()));

    let mut project = parse_and_compile(&source_tree).map_err(|e| e.composed(&source_tree))?;

    project.sources = source_tree;
    Ok(project)
}

fn parse_and_compile(source_tree: &SourceTree) -> Result<ProjectCompiled, ErrorMessages> {
    let options = Options::default()
        .with_target(Target::Sql(Some(Dialect::SQLite)))
        .no_format()
        .no_signature();

    // parse and resolve
    let ast_tree = prqlc::prql_to_pl_tree(source_tree)?;
    let mut root_module = semantic::resolve(ast_tree, Default::default())?;

    // find the database module
    let database_module = find_database_module(&mut root_module)?;

    // compile all main queries
    let mut queries = HashMap::new();
    let main_idents = root_module.find_mains();
    for main_ident in main_idents {
        let main_path: Vec<_> = main_ident.iter().cloned().collect();

        let rq;
        (rq, root_module) = semantic::lower_to_ir(root_module, &main_path, &database_module.path)?;
        let sql = prqlc::rq_to_sql(rq, &options)?;

        queries.insert(main_ident, sql);
    }
    Ok(ProjectCompiled {
        sources: SourceTree::default(), // placeholder
        queries,
        database_module,
        root_module,
    })
}

fn find_database_module(root_module: &mut RootModule) -> Result<DatabaseModule, Errors> {
    let lutra_sqlite = Ident::from_path(vec!["lutra", "sqlite"]);
    let db_modules_fq = root_module.find_by_annotation_name(&lutra_sqlite);

    let db_module_fq = match db_modules_fq.len() {
        0 => {
            return Err(Error::new_simple("cannot find the database module.")
                .push_hint("define a module annotated with `@lutra.sqlite`")
                .into());
        }
        1 => db_modules_fq.into_iter().next().unwrap(),
        _ => {
            return Err(Error::new_simple("cannot query multiple databases")
                .push_hint("you can define only one module annotated with `@lutra.sqlite`")
                .push_hint("this will be supported in the future")
                .into());
        }
    };

    // extract the declaration and retrieve its annotation
    let decl = root_module.module.get(&db_module_fq).unwrap();
    let annotation = decl
        .annotations
        .iter()
        .find(|x| prqlc::semantic::is_ident_or_func_call(&x.expr, &lutra_sqlite))
        .unwrap();

    let def_id = decl.declared_at;

    // make sure that there is exactly one arg
    let arg = match &annotation.expr.kind {
        prqlc::ir::pl::ExprKind::Ident(_) => {
            return Err(Error::new_simple("missing connection parameters")
                .push_hint("add `{file='sqlite.db'}`")
                .with_span(annotation.expr.span)
                .into());
        }
        prqlc::ir::pl::ExprKind::FuncCall(call) => {
            // TODO: maybe this should be checked by actual type-checker
            if call.args.len() != 1 {
                Err(Error::new_simple("expected exactly one argument")
                    .with_span(annotation.expr.span))?;
            }
            call.args.first().unwrap()
        }
        _ => unreachable!(),
    };

    let params = prqlc::semantic::static_eval(arg.clone(), root_module)?;
    let prqlc::ir::constant::ConstExprKind::Tuple(params) = params.kind else {
        return Err(Error::new_simple("expected exactly one argument")
            .with_span(params.span)
            .into());
    };

    let file = params.into_iter().next().unwrap();
    let prqlc::ir::constant::ConstExprKind::Literal(Literal::String(file_str)) = file.kind else {
        return Err(Error::new_simple("expected a string")
            .with_span(file.span)
            .into());
    };

    let file_relative = std::path::PathBuf::from_str(&file_str)
        .map_err(|e| Error::new_simple(e.to_string()).with_span(file.span))?;
    if !file_relative.is_relative() {
        Err(
            Error::new_simple("expected a relative path to the SQLite database file")
                .with_span(file.span),
        )?;
    }

    Ok(DatabaseModule {
        path: db_module_fq.into_iter().collect(),
        def_id,
        connection_params: SqliteConnectionParams { file_relative },
    })
}
