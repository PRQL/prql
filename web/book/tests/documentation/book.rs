#![cfg(not(target_family = "wasm"))]
use anyhow::{anyhow, bail, Result};
use globset::Glob;
use insta::assert_snapshot;
use itertools::Itertools;
use mdbook_prql::{code_block_lang_tags, LangTag};
use prqlc::{pl_to_prql, pl_to_rq, prql_to_pl};
use pulldown_cmark::Tag;
use std::fs;
use std::path::Path;
use walkdir::WalkDir;

use super::compile;

/// This test:
/// - Extracts PRQL code blocks from the book
/// - Compiles them to SQL, comparing to a snapshot.
/// - We raise an error if they shouldn't pass or shouldn't fail.
/// - Insta raises an error if there's a snapshot diff.
///
/// This mirrors the process in [`replace_examples`], which inserts a comparison
/// table of SQL into the book, and so serves as a snapshot test of those
/// examples.
//
// We re-use the code (somewhat copy-paste) for the other compile tests below.
#[test]
fn test_prql_examples_compile() -> Result<()> {
    let examples = collect_book_examples()?;

    let mut errs = Vec::new();
    for Example { name, tags, prql } in examples {
        let result = compile(&prql);
        let should_succeed = !tags.contains(&LangTag::Error);

        match (should_succeed, result) {
            (true, Err(e)) => errs.push(format!(
                "
---- {name} ---- ERROR
Use `prql error` as the language label to assert an error compiling the PRQL.

-- Original PRQL --
```
{prql}
```

-- Error --
```
{e}
```
"
            )),

            (false, Ok(output)) => errs.push(format!(
                "
---- {name} ---- UNEXPECTED SUCCESS
Succeeded compiling, but example was marked as `error`.
Remove `error` as a language label to assert successfully compiling.

-- Original PRQL --
```
{prql}
```

-- Result --
```
{output}
```
"
            )),
            (_, result) => {
                assert_snapshot!(name, result.unwrap_or_else(|e| e.to_string()), &prql);
            }
        }
    }
    if errs.is_empty() {
        Ok(())
    } else {
        Err(anyhow!(errs.join("\n")))
    }
}

#[test]
fn test_prql_examples_rq_serialize() -> Result<()> {
    for Example { tags, prql, .. } in collect_book_examples()? {
        // Don't assert that this fails, whether or not they compile to RQ is
        // undefined.
        if tags.contains(&LangTag::Error) {
            continue;
        }
        let rq = prql_to_pl(&prql).map(pl_to_rq)?;
        // Serialize
        serde_json::to_string(&rq).unwrap();
    }

    Ok(())
}

/// Test that the formatted result (the `Display` result) of each example can be
/// compiled.
//
// We previously snapshot all the queries. But that was a lot of output, for
// something we weren't yet looking at.
//
// The ideal would be to auto-format the examples themselves, likely during the
// compilation. For that to provide a good output, we need to implement a proper
// autoformatter.
#[test]
fn test_prql_examples_display_then_compile() -> Result<()> {
    let examples = collect_book_examples()?;

    let mut errs = Vec::new();
    for Example { name, tags, prql } in examples {
        let result = prql_to_pl(&prql)
            .and_then(|x| pl_to_prql(&x))
            .and_then(|x| compile(&x));

        let should_succeed = !tags.contains(&LangTag::NoFmt);

        match (should_succeed, result) {
            (true, Err(e)) => errs.push(format!(
                "
---- {name} ---- ERROR formatting & compiling
Use `prql no-fmt` as the language label to assert an error from formatting & compiling.

-- Original PRQL --

```
{prql}
```
-- Error --
```
{e}
```
"
            )),

            (false, Ok(output)) => errs.push(format!(
                "
---- {name} ---- UNEXPECTED SUCCESS after formatting
Succeeded at formatting and then compiling the prql, but example was marked as `no-fmt`.
Remove `no-fmt` as a language label to assert successfully compiling the formatted result.

-- Original PRQL --
```
{prql}
```
-- Result --
```
{output}
```
"
            )),
            _ => {}
        }
    }
    if errs.is_empty() {
        Ok(())
    } else {
        Err(anyhow!(errs.join("")))
    }
}

struct Example {
    /// Name contains the file, the heading, and the index of the example.
    name: String,
    tags: Vec<LangTag>,
    /// The PRQL text
    prql: String,
}

/// Collect all the PRQL examples in the book, as [Example]s.
/// Excludes any with a `no-eval` tag.
fn collect_book_examples() -> Result<Vec<Example>> {
    use pulldown_cmark::{Event, Parser};
    let glob = Glob::new("**/*.md")?.compile_matcher();
    Ok(WalkDir::new(Path::new("./src/"))
        .into_iter()
        .flatten()
        .filter(|x| glob.is_match(x.path()))
        .flat_map(|dir_entry| {
            let text = fs::read_to_string(dir_entry.path())?;
            // TODO: Still slightly duplicative logic here and in
            // [lib.rs/replace_examples], but not sure how to avoid it.
            //
            let mut parser = Parser::new(&text);
            let mut prql_blocks: Vec<Example> = vec![];
            // Keep track of the latest heading, so snapshots can have the
            // section they're in. This makes them easier to find and means
            // adding one example at the top of the book doesn't cause a huge
            // diff in the snapshots of that file's examples..
            let mut latest_heading = "".to_string();
            let file_name = &dir_entry
                .path()
                .strip_prefix("./src/")?
                .to_str()
                .unwrap()
                .trim_end_matches(".md");

            // Iterate through the markdown file, getting examples.
            while let Some(event) = parser.next() {
                if let Event::Start(Tag::Heading { .. }) = event.clone() {
                    if let Some(Event::Text(pulldown_cmark::CowStr::Borrowed(heading))) =
                        parser.next()
                    {
                        // We clear and then push because just setting
                        // `latest_heading` leads to lifetime issues.
                        latest_heading = heading
                            .chars()
                            .filter(|&c| c.is_ascii_alphanumeric() || c == '-' || c == ' ')
                            .collect();
                    }
                }
                let Some(tags) = code_block_lang_tags(&event) else {
                    continue;
                };

                if tags.contains(&LangTag::Prql) && !tags.contains(&LangTag::NoEval) {
                    let mut prql = String::new();
                    while let Some(Event::Text(line)) = parser.next() {
                        prql.push_str(line.to_string().as_str());
                    }
                    if prql.is_empty() {
                        bail!("Expected text in PRQL code block");
                    }
                    let heading = latest_heading.replace(' ', "-").to_ascii_lowercase();
                    // Only add the heading if it's different from the file name.
                    let name = if !file_name.ends_with(&heading) {
                        format!("{file_name}/{heading}")
                    } else {
                        file_name.to_string()
                    };
                    prql_blocks.push(Example { name, tags, prql });
                }
            }
            Ok(prql_blocks)
        })
        .flatten()
        // Add an index suffix to each path's examples (so we group by the path).
        .group_by(|e| e.name.clone())
        .into_iter()
        .flat_map(|(path, blocks)| {
            blocks.into_iter().enumerate().map(move |(i, e)| Example {
                name: format!("{path}/{i}"),
                ..e
            })
        })
        .collect())
}
