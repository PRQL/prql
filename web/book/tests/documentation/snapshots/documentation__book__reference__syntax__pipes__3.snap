---
source: web/book/tests/documentation/book.rs
expression: "let b =\n  \\ 3        # Line break OK within this line wrap\n\n# Line break OK before this `from` statement\n\nfrom db.y\nderive a = b\n"
---
SELECT
  *,
  3 AS a
FROM
  y
